File: elixir/collatz-conjecture/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  @initial_steps 0
  @terminating_number 1

  require Integer

  defguardp positive_integer?(input) when is_integer(input) and input > 0

  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(input) when positive_integer?(input), do: calc(input, @initial_steps)
  defp calc(@terminating_number, steps), do: steps

  defp calc(input, steps) when Integer.is_even(input) do
    input
    |> n_div_two()
    |> calc(steps + 1)
  end

  defp calc(input, steps) do
    input
    |> three_n_plus_one()
    |> calc(steps + 1)
  end

  defp n_div_two(n), do: div(n, 2)
  defp three_n_plus_one(n), do: 3 * n + 1
end

File: javascript/collatz-conjecture/collatz-conjecture.js
const INITIAL_STEPS = 0
const TERMINATING_NUMBER = 1
const IS_EVEN = number => number % 2 === 0
const N_DIV_TWO = n => n / 2
const THREE_N_PLUS_ONE = n => 3 * n + 1

export const steps = input => {
  if (input < 1) {
    throw new Error("Only positive numbers are allowed")
  }

  return calculateSteps(input, INITIAL_STEPS)
}

function calculateSteps(input, steps) {
  if (input === TERMINATING_NUMBER) {
    return steps
  }

  input = IS_EVEN(input) ? N_DIV_TWO(input) : THREE_N_PLUS_ONE(input)
  return calculateSteps(input, steps + 1)
}
